keycloak_setup/institution_list.py

```diff
@@ -28,371 +30,619 @@ class InstitutionMeta(TypedDict):
         "abbreviation": "UCLA",
         "is_US": True,
         "region": _US_CAN,
+        "authorlist": True,
+        "_ldap_o": "ucla",
     },
     "UCLouvain": {
         "cite": "Centre for Cosmology, Particle Physics and Phenomenology - CP3, Université catholique de Louvain, Louvain-la-Neuve, Belgium",
         "abbreviation": "UCLOUVAIN",
         "is_US": False,
         "region": _EUROPE,
+        "authorlist": True,
+        "_ldap_o": "uclouvain",
     },
     "Uppsala": {
         "cite": "Dept. of Physics and Astronomy, Uppsala University, Box 516, S-75120 Uppsala, Sweden",
         "abbreviation": "UU",
         "is_US": False,
         "region": _EUROPE,
+        "authorlist": True,
+        "_ldap_o": "uppsala",
     },
     "Utah": {
         "cite": "Department of Physics and Astronomy, University of Utah, Salt Lake City, UT 84112, USA",
         "abbreviation": "UTAH",
         "is_US": True,
         "region": _US_CAN,
+        "authorlist": True,
+        "_ldap_o": "utah",
     },
-    "UW-Madison-Astronomy": {
-        "cite": "Dept. of Astronomy, University of Wisconsin–Madison, Madison, WI 53706, USA",
-        "abbreviation": "UW",
-        "is_US": True,
-        "region": _US_CAN,
-    },
-    "UW-Madison-WIPAC": {
+    "UW-Madison": {
         "cite": "Dept. of Physics and Wisconsin IceCube Particle Astrophysics Center, University of Wisconsin–Madison, Madison, WI 53706, USA",
         "abbreviation": "UW",
         "is_US": True,
         "region": _US_CAN,
+        "authorlist": True,
+        "authorlists": {
+            "wipac": "Dept. of Physics and Wisconsin IceCube Particle Astrophysics Center, University of Wisconsin–Madison, Madison, WI 53706, USA",
+            "astro": "Dept. of Astronomy, University of Wisconsin–Madison, Madison, WI 53706, USA",
+        },
+        "_ldap_o": "uwmad",
     },
     "UW-River-Falls": {
         "cite": "Dept. of Physics, University of Wisconsin, River Falls, WI 54022, USA",
         "abbreviation": "UWRF",
         "is_US": True,
         "region": _US_CAN,
+        "authorlist": True,
+        "_ldap_o": "uwrf",
+    },
+    "Washington": {
+        "cite": "University of Washington",
+        "abbreviation": "UWa",
+        "is_US": True,
+        "region": _US_CAN,
+        "authorlist": False,
+        "_ldap_o": "washington",
     },
     "Wuppertal": {
         "cite": "Dept. of Physics, University of Wuppertal, D-42119 Wuppertal, Germany",
         "abbreviation": "WUPPERTAL",
         "is_US": False,
         "region": _EUROPE,
+        "authorlist": True,
+        "_ldap_o": "wuppertal",
     },
     "Yale": {
         "cite": "Dept. of Physics, Yale University, New Haven, CT 06520, USA",
         "abbreviation": "YALE",
         "is_US": True,
         "region": _US_CAN,
+        "authorlist": True,
+        "_ldap_o": "yale",
+    },
+}  # type: Dict[str, InstitutionMeta]
+
+GEN2_INSTS = {
```

Comment:
Is this dict for institutions that are only part of Gen 2? I would think this would make more sense as a flag on the other/main list than a separate entity.

Reply:
Yes.  I suppose I could also do a flag, but they are located in a separate spot in Keycloak.  Also, all regular IceCube institutions are added to Gen2.

Reply:
I think it would be simpler to have one large list, then filter down versus build up. That said as long as the "master list" is accessible from an interface, it doesn't much matter which way it's stored on the backend. I have the MOU Dash in mind, where "all the institutions" are available regardless of whether they have any SOWs, including Gen 2 institutions like Columbia.

Reply:
Will there be a way to get the "mast list"?

Reply:
One thing I should discuss with you is if you want an attr to mark MOU institutions, or if you have a whitelist on your side.

There will be a way to get a list of all groups, and then that can be filtered for institutions.  So, this could be a function inside MOU, a shared library, or we could make it into a proper REST API.  (this goes into my list of "things to clean up after Keycloak is ready")

Reply:
> I think it would be simpler to have one large list, then filter down versus build up.

By this I meant filter on the krs side, I would rather all the institution-subsetting logic be outsourced to krs. I'm a fan of asking (querying) for what you want on a REST API.

I don't see a reason for adding an MOU attr since it's not that picky. The "whitelist" is whether an institution leader adds SOWs--in other words, no whitelist/blacklist is needed. Let's just get all the institutions, whether it makes sense or not.

I'm not sure how you're building the routes, something like `/krs/institutions/all` would work (or `/krs/institutions/mou` if that doesn't work).

Reply:
Going to your comment on retired institutions, they still show up in Keycloak.  And there are a couple of them that don't have MoU agreements, they just have people with computing accounts.

Thus why I ask about an attr or whitelist.

Reply:
> Going to your comment on retired institutions, they still show up in Keycloak.

That's good. Are institutions ever deleted?

> And there are a couple of them that don't have MoU agreements, they just have people with computing accounts.
>
> Thus why I ask about an attr or whitelist.

Okay, sounds like filtering makes things easier. Let's add a `"has_mou"` bool flag, then send only these institutions on `/krs/institutions/mou` (or similar). How does that sound?

Reply:
> > Going to your comment on retired institutions, they still show up in Keycloak.
> 
> That's good. Are institutions ever deleted?

So far, no.  But maybe in the future at some point?  We can't delete users, but if no users are assigned to an institution then we can maybe delete it.

Note for the author list I've just made a flag/subgroups such that I can ignore some institutions.


> > And there are a couple of them that don't have MoU agreements, they just have people with computing accounts.
> > Thus why I ask about an attr or whitelist.
> 
> Okay, sounds like filtering makes things easier. Let's add a `"has_mou"` bool flag, then send only these institutions on `/krs/institutions/mou` (or similar). How does that sound?

I guess the question is if you want a nice REST route, which I'd have to make, or just a library function you can call which does the filtering?  Keycloak itself doesn't have anything like that, you just get a dump of all the groups.

Reply:
> > > Going to your comment on retired institutions, they still show up in Keycloak.
> > 
> > 
> > That's good. Are institutions ever deleted?
> 
> So far, no. But maybe in the future at some point? We can't delete users, but if no users are assigned to an institution then we can maybe delete it.
> 
> Note for the author list I've just made a flag/subgroups such that I can ignore some institutions.

I'm in favor of never deleting institutions, it would make historical comparisons easier for MOU data. Flags are great.

> > > And there are a couple of them that don't have MoU agreements, they just have people with computing accounts.
> > > Thus why I ask about an attr or whitelist.
> > 
> > 
> > Okay, sounds like filtering makes things easier. Let's add a `"has_mou"` bool flag, then send only these institutions on `/krs/institutions/mou` (or similar). How does that sound?
> 
> I guess the question is if you want a nice REST route, which I'd have to make, or just a library function you can call which does the filtering? Keycloak itself doesn't have anything like that, you just get a dump of all the groups.

It's not terrible if the client (MOU) has to filter if it's just a flag. Then again, if it's just a flag then could this be added to the krs interface, similar to a mongo projection?

Reply:
Resolved: `krs.institutions` will have client-accessible helper functions for filtering since the krs doesn't have a custom REST interface.